cmd/collector/app/span_processor.go

```diff
@@ -177,3 +209,79 @@ func (sp *spanProcessor) enqueueSpan(span *model.Span, originalFormat SpanFormat
 	}
 	return sp.queue.Produce(item)
 }
+
+func (sp *spanProcessor) StartStateReporting(reportPeriod time.Duration) {
+	sp.background(reportPeriod, sp.updateGauges)
+}
+
+func (sp *spanProcessor) StartDynQueueSizeUpdater(reportPeriod time.Duration) {
+	sp.background(reportPeriod, sp.updateQueueSize)
+}
+
+func (sp *spanProcessor) background(reportPeriod time.Duration, callback func()) {
+	go func() {
+		ticker := time.NewTicker(reportPeriod)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				callback()
+			case <-sp.stopCh:
+				return
+			}
+		}
+	}()
+}
+
+func (sp *spanProcessor) updateQueueSize() {
+	if sp.spansProcessed.Load() == 0 {
+		return
+	}
+
+	if sp.dynQueueSizeWarmup == 0 {
+		return
+	}
+
+	if sp.dynQueueSizeMemory == 0 {
+		sp.logger.Warn("The dynamic queue size warmup value is set, but not the amount of memory to use. Skipping.")
+		return
+	}
+
+	if sp.spansProcessed.Load() < uint64(sp.dynQueueSizeWarmup) {
+		return
+	}
+
+	// first, we get the average size of a span, by dividing the bytes processed by num of spans
+	average := sp.bytesProcessed.Load() / sp.spansProcessed.Load()
+
+	// finally, we divide the available memory by the average size of a span
+	idealQueueSize := float64(sp.dynQueueSizeMemory / uint(average))
+
+	// cap the queue size, just to be safe...
+	if idealQueueSize > maxQueueSize {
+		idealQueueSize = maxQueueSize
+	}
+
+	var diff float64
+	current := float64(sp.queue.Capacity())
+	if idealQueueSize > current {
+		diff = idealQueueSize / current
+	} else {
+		diff = current / idealQueueSize
+	}
+
+	// resizing is a costly operation, we only perform it if we are at least 10% apart from the desired value
```

Comment:
10% seems pretty low tolerance

Reply:
What would be a better value? 20%? Note that this is only calculated after the warmup phase. In any case, I increased it to 20%. We can then see how it works in the real world and adjust accordingly.